<a href="https://colab.research.google.com/github/Aingon14/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [120]:
import pandas as pd
import os

In [121]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/bsc_dpdm23_data'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [122]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

In [123]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##quiz : ข้อมูลมาจากกี่ประเทศ

In [124]:
len(set(data['Country']))

38

In [125]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

# แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย groupby.apply(list)

In [126]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [127]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [11]:
!pip install apyori

In [12]:
from apyori import apriori

In [13]:
list(apriori(transactions,min_support= 0.065))   #คนที่ซื้อของซ้ำๆกัน

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [14]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [15]:
len(frequentItemsets)

4

In [16]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [17]:
frequentItemsets[0][0]

frozenset({'22423'})

In [18]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [19]:
frequentItemsets[0][0].__len__()

1

In [20]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [21]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

#Quiz 22386 และ 85099B คือสินค้าอะไร

In [53]:
# 22386, 85099B คือสินค้อะไร
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [54]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

In [24]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'568060',
 '546683',
 '557064',
 '575490',
 '550470',
 '546442',
 '576063',
 '580673',
 '568676',
 '543188',
 '571797',
 'C567358',
 '562711',
 '566094',
 '543019',
 '565917',
 '555477',
 '556459',
 '550953',
 '562420',
 '547580',
 '553535',
 '558923',
 '580053',
 '573397',
 '574076',
 'C550648',
 '565234',
 '578936',
 '570720',
 '564541',
 '570677',
 '562033',
 '557502',
 '548648',
 '579512',
 '540464',
 '564972',
 '553210',
 '544089',
 '561095',
 '547249',
 '555400',
 '557262',
 '548373',
 '549016',
 '552308',
 '577498',
 '554054',
 '575739',
 '540005',
 '560976',
 '554643',
 '536946',
 '559491',
 '567293',
 '574298',
 '558103',
 '561051',
 '548919',
 '560246',
 '569834',
 'C552016',
 '559295',
 '547826',
 '550915',
 '549730',
 '549425',
 '537022',
 '543036',
 '544434',
 '563534',
 '565916',
 '570270',
 '536876',
 'C545449',
 '551995',
 '542102',
 'C558255',
 '562219',
 '544681',
 '536980',
 '560634',
 '552816',
 '580665',
 '560895',
 '541870',
 '550205',
 '557431',
 '542344',
 '577

In [25]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

# HW1 หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลาตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน (groupby ประเทศกับเดือน)

In [91]:
data['InvoiceDate'][0]

'12/01/2010 08:26'

In [92]:
data['InvoiceDate'][0].split()[0] #ตัด

'12/01/2010'

In [93]:
date =[]
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)

In [95]:
date[0].split('/')

['12', '01', '2010']

In [111]:
#แยกวัน เดือน ปี ออกจากกัน
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  m = date[i].split('/')[0]
  d = date[i].split('/')[1]
  y = date[i].split('/')[2]
  month.append(m)
  day.append(d)
  year.append(y)

In [112]:
data_test=data

In [113]:
data_test['date']=date
data_test['day']=day
data_test['month']=month
data_test['year']=year

In [114]:
#แปลงวัน
x=[]
for i in range(1,32):
  x.append(i)
print(x)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]


In [115]:
data_day=data_test.replace(['01', '02', '03', '04', '05', '06', '07', '08', '09', '10','11',
                                      '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31'],x)   #แทนที่วันจาก str

In [116]:
data_day

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

        day month  year  
0         1    12  2010  
1         1    12  2010  
2         1    12  2010  
3         1    12  2010  
4         1    12  2010  
...     ...   ...   ...  
541904    9    12  2011  
541905    9    12  2011  
541906    9    12  2011  
541907    9    12  2011  
541908    9    12  2011  

[541909 rows x 12 columns]

In [117]:
set(data_test['month'])  #ข้อมูลมีทั้ง str และแบบตัวเลข

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '1',
 '10',
 '11',
 '12',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9'}

In [118]:
data_test1=data_test.replace(to_replace=["1", "2",'3','4','5','6','7','8','9'],value=[1,2,3,4,5,6,7,8,9])
data_test1

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
541904  09    12  2011  
541905  09    12  2011  
541906  09    12  2011  
541907  09    12  2011  
541908  09    12  2011  

[541909 rows x 12 columns]

In [119]:
data_test.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
date           False
day            False
month          False
year           False
dtype: bool

In [104]:
data_test

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
541904  09    12  2011  
541905  09    12  2011  
541906  09    12  2011  
541907  09    12  2011  
541908  09    12  2011  

[541909 rows x 12 columns]

## ข้อ1


In [105]:
data_C_N= data_test[data_test['month']==12]
data_C_N

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country, date, day, month, year]
Index: []

In [106]:
!pip install apyori
from apyori import apriori

In [107]:
transactions_C_N = data_C_N.groupby('InvoiceNo')['StockCode'].apply(list)
transactions_C_N

<ipython-input-107-e686f91dd3a2>:1: FutureWarning: Not prepending group keys to the result index of transform-like apply. In the future, the group keys will be included in the index, regardless of whether the applied function returns a like-indexed object.
To preserve the previous behavior, use

	>>> .groupby(..., group_keys=False)

To adopt the future behavior and silence this warning, use 

	>>> .groupby(..., group_keys=True)
  transactions_C_N = data_C_N.groupby('InvoiceNo')['StockCode'].apply(list)


Series([], Name: StockCode, dtype: object)

In [43]:
#frequentItemsets = list(apriori(transactions_C_N,min_support = 0.05))
#frequentItemsets

In [108]:
frequentItemsets_C_N = list(apriori(transactions_C_N,min_support = 0.04))
for i in frequentItemsets_C_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)

## group by  ข้อ1

In [109]:
transactions_C_N = data_C_N.groupby(['Country', 'month'])['StockCode'].apply(list)

<ipython-input-109-0dc4b228102a>:1: FutureWarning: Not prepending group keys to the result index of transform-like apply. In the future, the group keys will be included in the index, regardless of whether the applied function returns a like-indexed object.
To preserve the previous behavior, use

	>>> .groupby(..., group_keys=False)

To adopt the future behavior and silence this warning, use 

	>>> .groupby(..., group_keys=True)
  transactions_C_N = data_C_N.groupby(['Country', 'month'])['StockCode'].apply(list)


In [110]:
transactions_C_N

Series([], Name: StockCode, dtype: object)

In [1]:
frequentItemsets_C_N = list(apriori(transactions_C_N,min_support = 0.03))
for i in frequentItemsets_C_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)

NameError: name 'apriori' is not defined

In [128]:
import datetime

In [129]:
data['InvoiceDate']

0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [132]:
data['InvoiceDate']=pd.to_datetime(data.InvoiceDate, format='%m/%d/%Y')

ValueError: unconverted data remains:  08:26

In [130]:
data['Date'] = data['InvoiceDate'].dt.date

AttributeError: Can only use .dt accessor with datetimelike values